src/rsz/src/Resizer.cc

```diff
@@ -626,33 +627,92 @@ void Resizer::getPins(Instance* inst, PinVector& pins) const
+  odb::dbModNet* to_mod_net = iterm_to->getModNet();
+
+  odb::dbModNet* from_mod_net = iterm_from->getModNet();
+  dbNet* from_db_net = iterm_from->getNet();
+
+  std::string required_name
+      = from_mod_net ? from_mod_net->getName() : from_db_net->getName();
+  std::string to_name
+      = to_mod_net ? to_mod_net->getName() : to_db_net->getName();
+
+  if (from_mod_net && to_mod_net) {
+    from_mod_net->rename(to_name.c_str());
+    to_mod_net->rename(required_name.c_str());
+  } else if (from_db_net && to_db_net) {
+    to_db_net->swapNetNames(from_db_net);
+  } else if (from_mod_net && to_db_net) {
+    to_db_net->rename(required_name.c_str());
+    from_mod_net->rename(to_name.c_str());
+  } else if (to_mod_net && from_db_net) {
+    to_mod_net->rename(required_name.c_str());
+    from_db_net->rename(to_name.c_str());
+  }
+}
+
 Instance* Resizer::bufferInput(const Pin* top_pin, LibertyCell* buffer_cell)
 {
-  Term* term = db_network_->term(top_pin);
-  Net* input_net = db_network_->net(term);
+  odb::dbITerm* top_pin_ip_iterm;
+  odb::dbBTerm* top_pin_ip_bterm;
+  odb::dbModITerm* top_pin_ip_moditerm;
+  odb::dbModBTerm* top_pin_ip_modbterm;
+
+  db_network_->staToDb(top_pin,
+                       top_pin_ip_iterm,
+                       top_pin_ip_bterm,
+                       top_pin_ip_moditerm,
+                       top_pin_ip_modbterm);
+
+  Net* input_net = nullptr;
+  dbNet* top_pin_flat_net = top_pin_ip_bterm->getNet();
+  odb::dbModNet* top_pin_hier_net = top_pin_ip_bterm->getModNet();
+
+  if (top_pin_hier_net) {
+    input_net = db_network_->dbToSta(top_pin_hier_net);
+  } else if (top_pin_flat_net) {
+    input_net = db_network_->dbToSta(top_pin_flat_net);
+  }
+
   LibertyPort *input, *output;
   buffer_cell->bufferPorts(input, output);
 
   bool has_non_buffer = false;
   bool has_dont_touch = false;
+
   NetConnectedPinIterator* pin_iter = network_->connectedPinIterator(input_net);
   while (pin_iter->hasNext()) {
     const Pin* pin = pin_iter->next();
     // Leave input port pin connected to input_net.
     if (pin != top_pin) {
       auto inst = network_->instance(pin);
+      odb::dbInst* db_inst;
+      odb::dbModInst* mod_inst;
+      db_network_->staToDb(inst, db_inst, mod_inst);
       if (dontTouch(inst)) {
         has_dont_touch = true;
         logger_->warn(RSZ,
                       85,
                       "Input {} can't be buffered due to dont-touch fanout {}",
-                      network_->name(input_net),
+                      (input_net ? network_->name(input_net) : "no-input-net"),
                       network_->name(pin));
         break;
       }
       auto cell = network_->cell(inst);
-      auto lib = network_->libertyCell(cell);
-      if (lib && !lib->isBuffer()) {
+      if (db_inst) {
+        auto lib = network_->libertyCell(cell);
+        if (lib && !lib->isBuffer()) {
+          has_non_buffer = true;
+        }
+      } else {
         has_non_buffer = true;
```

Comment:
Is it correct that a modinst is counted as a non-buffer?

Reply:
Another good thought.

It turns out that when getting all the net connected pins the iterator will indeed see the buffer (if any) inside the modinst (from the flat net traversal part). It will also see the moditerm in the traversal on the modinst (as part of the modnet part of the traversal).

So it will catch if there is a buffer in the immediate fanout of the top pin even if that fanout is a modinst with the moditerm connected to the iterm of a buffer. 

This is a case where the dual model really helps.

Reply:
Yes, the code handles modinsts correctly